Feature/68 add fixed time scheduling

kevin-wijnen opened the pull request:
Yet to merge

kevin-wijnen committed:
Experiment single scheduled task

kevin-wijnen committed:
Structural/Design commentary

kevin-wijnen committed:
Fixed experiment with multiple tasks

kevin-wijnen committed:
Several more experiments added

- SingleTask fixed for execution during JUnit
- Added checkDoneStatus to check whether tasks have been successfully
executed
- Migrated from newSingleThreadScheduledExecutor to
ScheduledThreadPoolExecutor(1) to get access to the Purge function (to
prevent memory leaks and clear the task queue)

kevin-wijnen committed:
Deletion of unnecessary SwitchingMomentCalculator

kevin-wijnen committed:
Implementation Fixed Time Scheduler

kevin-wijnen committed:
General clean up + added .purge command

kevin-wijnen committed:
Merge branch 'master' into feature/68-add-fixed-time-scheduling

# Conflicts:
#	iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/EventDataListener.java
#	iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/Schedule.java
#	iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/SwitchingMoment.java
#	iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculator.java
#	iec61850serversimulator/src/test/java/com/cgi/iec61850serversimulator/SchedulerTest.java
#	iec61850serversimulator/src/test/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculatorTest.java

kevin-wijnen committed:
SwitchingMomentCalculator fixed

kevin-wijnen committed:
Switching Moment calculation implementation in Scheduler

kevin-wijnen committed:
TimeCalculator unit test made + code layout tweaks

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/EventDataListener.java:
```diff
@@ -340,7 +331,12 @@ public void dailySwitchingMomentCalculation() {
+				this.scheduler.calculateTasks(this.device);
+			} catch (SwitchingMomentCalculationException e) {
+				// TODO Auto-generated catch block
+				e.printStackTrace();
```
Replace this by a logger.warn statement.

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/Scheduler.java:
```diff
@@ -10,6 +14,24 @@
+	private Device device;
+
+	public Scheduler(Device device) {
+		ScheduledThreadPoolExecutor executor = (ScheduledThreadPoolExecutor) Executors.newScheduledThreadPool(1);
```
Replace all tabs by four spaces and configure the IDE to use four spaces instead of tabs.

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/ServerSimulator.java:
```diff
@@ -106,7 +106,7 @@ public void run() {
-			edl.getScheduler().switchingMomentCalculation(device);
+			edl.getScheduler().calculateTasks(device);
 		} catch (Exception e) {
 			logger.info("Initial switching moment calculation failed, try sending another schedule.");
```
An exception should be at either WARN or ERROR level.

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/ServerSimulator.java:
```diff
@@ -106,7 +106,7 @@ public void run() {
 		// Initial schedule
 		try {
-			edl.getScheduler().switchingMomentCalculation(device);
+			edl.getScheduler().calculateTasks(device);
```
Isn't the scheduler object from line 100 usable?

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/TimeCalculator.java:
```diff
@@ -0,0 +1,14 @@
+import java.time.LocalDateTime;
+
+public class TimeCalculator {
+	public int calculateRelativeTime(SwitchingMoment switchingMoment, LocalDateTime scheduledTime) {
```
- Make the method static
- Don't pass more information than needed: you only need the triggerTime from switchingMoment, so pass in "LocalDateTime triggerTime" instead of the full SwitchingMoment object. After changing this, you can make the "method signature" more meaningful. For instance: public static int calculateSecondsUntil(LocalDateTime from, LocalDateTime until).

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/test/java/com/cgi/experiments/SchedulerExperiment.java:
```diff
@@ -0,0 +1,149 @@
+package com.cgi.experiments;
```
I'll ignore the code from the experiments package in my review as it only contains personal experiments.

rlemmers commented on iec61850serversimulator/src/test/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculatorTest.java:
```diff
@@ -67,8 +67,8 @@ public void checkMockSchedule() {
-				.fixedTimeInt(fixedTimeInt).fixedTimeOn(fixedTimeOn).fixedTimeOff(fixedTimeOff)
-				.timeOn(timeOn).timeOff(timeOff).burningMins(burningMinutes).isEnabled(enabled).buildSchedule();
+				.fixedTimeInt(fixedTimeInt).fixedTimeOn(fixedTimeOn).fixedTimeOff(fixedTimeOff).timeOn(timeOn)
+				.timeOff(timeOff).burningMins(burningMinutes).isEnabled(enabled).buildSchedule();
```
The code seems to create almost the exact same schedule three times.
Suggestion: move the parts that are exactly the same to a separate method. You can return a ScheduleBuilder from that method, so you can add/change the remaining one or two fields later.

kevin-wijnen replied:
You did it

rlemmers commented on iec61850serversimulator/src/test/java/com/cgi/iec61850serversimulator/TimeCalculatorTest.java:
```diff
@@ -0,0 +1,34 @@
+public class TimeCalculatorTest {
+	/**
+	 * JUnit test class which tests the time calculation function.
+	 */
```
Looks like JavaDoc for the class: put it directly above the line "public class TimeCalculatorTest".

kevin-wijnen replied:
Done

kevin-wijnen committed:
Null Executor, Device and ScheduledFutures fix

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/ServerSimulator.java:
```diff
@@ -109,6 +109,7 @@ public void run() {
 			edl.getScheduler().calculateTasks(device);
 		} catch (Exception e) {
 			logger.info("Initial switching moment calculation failed, try sending another schedule.");
+			e.printStackTrace();
```
Never include printStackTrace, always use logger.warn/error("some message", e)

kevin-wijnen replied:
Done

kevin-wijnen committed:
Duplicate schedule fix, BROKEN CODE!

kevin-wijnen committed:
Initializing schedule crash fix

kevin-wijnen committed:
Feedback integrated

- Logger.warn replacements
- Space replacements
- TimeCalculator's function made static and integrated said static call
- JavaDoc fixed
- printStackTrace replacements

rlemmers committed:
68 ~ Simplifies creation of a Schedule

kevin-wijnen committed:
Scheduler reference fixed

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/Relay.java:
```diff
@@ -59,8 +61,12 @@ public void setSchedules(Schedule[] schedules) {
 	}
 
 	public void initializeSchedules(ModelNode scheduleInfo, int relayNr) {
+		String SWITCH_ROOT = "SWDeviceGenericIO/XSWC";
```
Move this to the top of the class and make it "public static final String ...", to mark it as a constant.

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/Scheduler.java:
```diff
@@ -15,98 +16,106 @@
-	public Scheduler(Device device) {
-		ScheduledThreadPoolExecutor executor = (ScheduledThreadPoolExecutor) Executors.newScheduledThreadPool(1);
+    public Scheduler(Device device) {
+        ScheduledThreadPoolExecutor executor = (ScheduledThreadPoolExecutor) Executors.newScheduledThreadPool(1);
```
Easier: this.executor = (ScheduledThreadPoolExecutor) Executors.newScheduledThreadPool(1);
If you do so, you can delete line 35.

kevin-wijnen replied:
Done both.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/Scheduler.java:
```diff
@@ -15,98 +16,106 @@
 
-	// Base it on feature/example-scheduling code! Triggering at certain
-	// times instead of a continuous check!
+        List<ScheduledFuture<?>> scheduledFutures = new ArrayList<ScheduledFuture<?>>();
+        this.scheduledFutures = scheduledFutures;
```
Easier: this.scheduledFutures = new ArrayList<ScheduledFuture<?>>();

kevin-wijnen replied:
Done.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/Scheduler.java:
```diff
@@ -15,98 +16,106 @@
-			this.executor.purge();
-		}
-	}
+    private void clearFuture(List<ScheduledFuture<?>> scheduledFutures) {
```
Clearer name: clearScheduledFutures (and I would even call it clearScheduledSwitchingMoments).

kevin-wijnen replied:
Done, also renamed the variable to scheduledSwitchingMoments to make the variable name also clear.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/TimeCalculator.java:
```diff
@@ -4,11 +4,10 @@
-		LocalDateTime triggerTime = switchingMoment.getTriggerTime();
-		Duration relativeTime = Duration.between(scheduledTime, triggerTime);
+    public static int calculateSecondsUntil(LocalDateTime from, LocalDateTime until) {
+        Duration relativeTime = Duration.between(until, from);
```
This calculates the time from "until" to "from". It means you've switched those two arguments.
In the call to this method, the currentTime is given as second argument. currentTime is the "fromTime", so it should be the first argument.

kevin-wijnen replied:
Done, along with migrating the changes to the method calls.

rlemmers commented on iec61850serversimulator/src/test/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculatorTest.java:
```diff
@@ -16,275 +16,242 @@
-
-		return schedule;
-	}
+    /**
```
Put the JavaDoc just about the "public class ...." line.

kevin-wijnen replied:
Done.

rlemmers commented on iec61850serversimulator/src/test/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculatorTest.java:
```diff
@@ -16,275 +16,242 @@
+    private static final Logger logger = LoggerFactory.getLogger(SwitchingMomentCalculatorTest.class);
+
+    // Initializing test Device
+    Device device = new Device();
```
Mark this variable as private.

kevin-wijnen replied:
Done.

rlemmers commented on iec61850serversimulator/src/test/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculatorTest.java:
```diff
@@ -16,275 +16,242 @@
+
+        final Schedule schedule = this.builderForDefaultSchedule(scheduleNr).buildSchedule();
+
+        // Schedule schedule = this.getMockSchedule(scheduleNr - 1, relayNr,
+        // dayInt,
+        // fixedTimeInt, fixedTimeOn,
+        // fixedTimeOff, timeOn, timeOff, burningMinutes);
```
Remove this commented code.

kevin-wijnen replied:
Done.

kevin-wijnen committed:
Feedback integrated

- General code/comment formatting improved
- Executor and ScheduledSwitchingMoments initialization simplified
- Clearer names for variables and method regarding scheduled switching
moments
- Switched the order of the Time Calculator's method call variables to
make more sense
- Removed unnecessary variables, commented out code

rlemmers reviewed (approved):
👍
